src/models/application.rs

```diff
@@ -0,0 +1,379 @@
+use crate::build_platform::Build;
+use crate::cloud_provider::models::{EnvironmentVariable, Storage};
+use crate::cloud_provider::service::{
+    delete_stateless_service, scale_down_application, Application as ApplicationService,
+};
+use crate::cloud_provider::service::{
+    deploy_stateless_service_error, deploy_user_stateless_service, send_progress_on_long_task, Action, Create, Delete,
+    Helm, Pause, Service, ServiceType, StatelessService,
+};
+use crate::cloud_provider::utilities::{print_action, sanitize_name};
+use crate::cloud_provider::DeploymentTarget;
+use crate::cmd::kubectl::ScalingKind::{Deployment, Statefulset};
+use crate::errors::EngineError;
+use crate::events::{EnvironmentStep, EventDetails, Stage, ToTransmitter, Transmitter};
+use crate::io_models::{Context, Listen, Listener, Listeners, Port, QoveryIdentifier};
+use crate::logger::Logger;
+use crate::models::types::CloudProvider;
+use function_name::named;
+use std::marker::PhantomData;
+
+#[derive(thiserror::Error, Debug)]
+pub enum ApplicationError {
+    #[error("Application invalid configuration: {0}")]
+    InvalidConfig(String),
+}
+
+pub struct Application<T: CloudProvider> {
+    _marker: PhantomData<T>,
+    pub(crate) context: Context,
```

Comment:
OPT: Can we avoid exposing internal fields to the outside world, making those private by default?

Reply:
it is pub(crate) so it is only exposed inside the models/ directory/crate, not outside

Reply:
Sure, but why need to expose it even to models/ directory/crate?

Reply:
So that each cloud provider can use them, if you are in the implementation Application<AWS> if it is not in pub(crate) you can't access self.port for example

Reply:
ok ok